### PR TITLE
Feat ⚙️ / Add planned stake/unstake estimates to the API

### PR DESCRIPTION
**Pull request by butonium**
Fix: #12

### CODE REVIEW

File: store/src/utils.rs [hunk 1]
```diff
@@ -610,6 +610,61 @@ pub async fn load_validators(
...
+                target_stake_mnde,
+                target_stake_msol,
+                scoring_run_id
+            FROM scores
+            WHERE eligible_stake_algo = true AND scoring_run_id IN (SELECT MAX(scoring_run_id) FROM scores) ORDER BY rank",
+            &[],
+        )
+        .await?;
+
+    let records: HashMap<_, _> = tokio::task::spawn_blocking(move || {
```
Comment (janlegner) [inline store/src/utils.rs:637]: the blocking thread is not needed here

File: api/src/handlers/reports_staking.rs [hunk 2]
```diff
@@ -1,18 +1,91 @@
...
+    });
+    Ok(records)
+}
+
+
+pub async fn handler(context: WrappedContext) -> Result<impl Reply, warp::Rejection> {
+    info!("Serving the staking report");
+    metrics::REQUEST_COUNT_REPORT_STAKING.inc();
+    let mut stakes: Vec<Stake> = Vec::new();
+    if let Ok(planned_stakes) = get_planned_stakes(context).await {
```
Comment (janlegner) [inline api/src/handlers/reports_staking.rs:73]: unnest + log in the else branch some error (and respond with an error)

File: api/src/handlers/reports_staking.rs [hunk 3]
```diff
@@ -1,18 +1,91 @@
...
+        Some(last_epoch) => last_epoch,
+        _ => return Ok(Default::default()),
+    };
+
+    for current_validator in validators.values().into_iter() {
+        if let Some(current_epoch_stats) = current_validator.epoch_stats.iter().filter(|validator| validator.epoch == last_epoch).last() {
+            let marinade_stake = current_epoch_stats.marinade_stake as u128;
+            if let Some(validator_score) = scores.get(&current_validator.vote_account) {
+                let should_have = (validator_score.target_stake_algo + validator_score.target_stake_mnde + validator_score.target_stake_msol) as u128;
+                if validator_score.eligible_stake_algo && (should_have as f64 - marinade_stake as f64 != 0.0) {
```
Comment (janlegner) [inline api/src/handlers/reports_staking.rs:47]: unnest here as well please

File: api/src/handlers/reports_staking.rs [hunk 4]
```diff
@@ -1,18 +1,91 @@
...
+                    records.push(StakingChange {
+                        identity: current_validator.identity.clone(),
+                        score: validator_score.score,
+                        current_stake: marinade_stake,
+                        next_stake: should_have * 1_000_000_000,
```
Comment (janlegner) [inline api/src/handlers/reports_staking.rs:52]: this can be in sol, maybe let's use `next_stake_sol`

File: api/src/handlers/reports_staking.rs [hunk 5]
```diff
@@ -1,18 +1,91 @@
...
+                    })
+                }
+            }
+        }
+    }
+    records.sort_by_key(|a| {
```
Comment (janlegner) [inline api/src/handlers/reports_staking.rs:58]: move to a separate function please

File: store/src/utils.rs [hunk 6]
```diff
@@ -611,6 +611,61 @@ pub async fn load_validators(
...
+                ui_hints,
+                eligible_stake_algo,
+                eligible_stake_mnde,
+                eligible_stake_msol,
+                target_stake_algo,
+                target_stake_mnde,
+                target_stake_msol,
+                scoring_run_id
+            FROM scores
+            WHERE eligible_stake_algo = true AND scoring_run_id IN (SELECT MAX(scoring_run_id) FROM scores) ORDER BY rank",
```
Comment (janlegner) [inline store/src/utils.rs:633]: why only `eligible_stake_algo = true`?

Comment (butonium) [reply store/src/utils.rs:633]: removed

**Comment by butonium** [comment]
In the latest commit I fixed a bug regarding validators with multiple identities and also improved the response time by adding current stakes in cache.

**Review by janlegner — APPROVED** [review]
let's merge and deploy this and let's remove the extra cache + improve performance after the cleanup task we discussed on the standup

**Pull request merged**